config/config_params.go

```diff
@@ -356,45 +356,50 @@ func (config *Config) resolve() (changed bool, err error) {
 	return
 }
 
-func (config *Config) DatastoreConfig() apiconfig.CalicoAPIConfig {
-	// Special case for etcdv2 datastore, where we want to honour established Felix-specific
-	// config mechanisms.
-	if config.DatastoreType == "etcdv2" {
-		// Build a CalicoAPIConfig with the etcd fields filled in from Felix-specific
-		// config.
-		var etcdEndpoints string
-		if len(config.EtcdEndpoints) == 0 {
-			etcdEndpoints = config.EtcdScheme + "://" + config.EtcdAddr
-		} else {
-			etcdEndpoints = strings.Join(config.EtcdEndpoints, ",")
-		}
-		etcdCfg := apiconfig.EtcdConfig{
-			EtcdEndpoints:  etcdEndpoints,
-			EtcdKeyFile:    config.EtcdKeyFile,
-			EtcdCertFile:   config.EtcdCertFile,
-			EtcdCACertFile: config.EtcdCaFile,
-		}
-		return apiconfig.CalicoAPIConfig{
-			Spec: apiconfig.CalicoAPIConfigSpec{
-				DatastoreType: apiconfig.EtcdV2,
-				EtcdConfig:    etcdCfg,
-			},
-		}
-	}
+func (config *Config) setBy(name string, source Source) bool {
+	_, set := config.sourceToRawConfig[source][name]
+	return set
+}
 
-	// Build CalicoAPIConfig from the environment.  This means that any XxxYyy field in
-	// CalicoAPIConfigSpec can be set by a corresponding XXX_YYY or CALICO_XXX_YYY environment
-	// variable, and that the datastore type can be set by a DATASTORE_TYPE or
-	// CALICO_DATASTORE_TYPE variable.  (Except in the etcdv2 case which is handled specially
-	// above.)
+func (config *Config) explicitlySet(name string) bool {
+	return config.setBy(name, ConfigFile) || config.setBy(name, EnvironmentVariable)
+}
+
+func (config *Config) DatastoreConfig() apiconfig.CalicoAPIConfig {
+	// We want Felix's datastore connection to be fully configurable using the same
+	// CALICO_XXX_YYY (or just XXX_YYY) environment variables that work for any libcalico-go
+	// client - for both the etcdv3 and KDD cases.  However, for the etcd case, Felix has for a
+	// long time supported FELIX_XXXYYY environment variables, and we want those to keep working
+	// too.
+
+	// To achieve that, first build a CalicoAPIConfig using libcalico-go's
+	// LoadClientConfigFromEnvironment - which means incorporating defaults and CALICO_XXX_YYY
+	// and XXX_YYY variables.
 	cfg, err := apiconfig.LoadClientConfigFromEnvironment()
 	if err != nil {
 		log.WithError(err).Panic("Failed to create datastore config")
 	}
-	// If that didn't set the datastore type (in which case the field will have been set to its
-	// default 'etcdv2' value), copy it from the Felix config.
-	if cfg.Spec.DatastoreType == "etcdv2" {
-		cfg.Spec.DatastoreType = apiconfig.DatastoreType(config.DatastoreType)
+
+	// Now allow FELIX_XXXYYY variables to override that, when they are explicitly set, in the
+	// etcd case.
+	if config.explicitlySet("DatastoreType") && config.DatastoreType == "etcdv3" {
```

Comment:
Any chance we'll reintroduce etcdv2 support as we head to GA?  Maybe we should match on etcdv2 or etcdv3?

Reply:
I'd be happy to do that if we also still had 'etcdv2' as a valid value for DatastoreType.  I think we should leave adding `|| config.DatastoreType == 'etcdv2'` here until the same time as we reinstate 'etcdv2' as a valid value - and I think we would be naturally prompted for that, if we also reinstated any etcdv2 testing, or added etcdv2 with similar config as we have for etcdv3.  WDYT?

Reply:
SGTM